File: scala/functionalcolorselector/src/main/scala/Gui.scala
import java.awt.{Color, Dimension}

import cats.effect.IO
import javax.swing.JFrame

import scala.util.Random

object Gui {

  private def randomColor(): IO[Color] =
    for {
      f <- IO.pure(() => Random nextInt 255)
      r <- IO.apply(f())
      g <- IO.apply(f())
      b <- IO.apply(f())
    } yield new Color(r,g,b)

  private def initFrameProps(f: JFrame, p: ColorChangerPanel): IO[Unit] =
    for {
      _ <- IO.apply(f.getContentPane.add(p))
      _ <- IO.apply(f.addKeyListener(p))
      _ <- IO.apply(f.setDefaultCloseOperation(3)) // EXIT_ON_CLOSE
      _ <- IO.apply(f.pack())
      _ <- IO.apply(f.setVisible(true))
      _ <- IO.apply(f.setFocusable(true))
      _ <- IO.apply(f.setResizable(false))
      _ <- IO.apply(f.validate())
    } yield ()

  def updatePanelBackgroundColor(p: ColorChangerPanel): IO[Unit] =
    for {
      c <- randomColor()
      _ <- IO.apply(p setBackground c)
      _ <- IO.apply(p.repaint())
    } yield ()

  def initPanelAndWaitForKeyPress(width: Int, height: Int): IO[Unit] =
    for {
      f <- IO.pure(new JFrame("Press SPACE key to update background color"))
      d <- IO.pure(new Dimension(width, height))
      p <- IO.pure(new ColorChangerPanel)
      _ <- IO.apply(p setPreferredSize d)
      _ <- initFrameProps(f, p)
    } yield ()

}

File: kotlin/kotlinFunctionalColorSelector/src/main/kotlin/functionalcolorselector/Gui.kt
package functionalcolorselector

import arrow.fx.IO
import java.awt.Color
import java.awt.Dimension
import javax.swing.JFrame
import kotlin.random.Random

object Gui {

    private fun randomColor(): IO<Color> {
        return IO { { Random.nextInt(255) } }.flatMap { f ->
            IO { f() }.flatMap { r ->
                IO { f() }.flatMap { g ->
                    IO { f() }.map { b ->
                        Color(r, g, b)
                    }
                }
            }
        }
    }

    private fun initFrameProps(f: JFrame, p: ColorChangerPanel): IO<Unit> {
        return IO { f.contentPane.add(p) }
            .flatMap { IO { f.addKeyListener(p) } }
            .flatMap { IO { f.setDefaultCloseOperation(3) } }
            .flatMap { IO { f.pack() } }
            .flatMap { IO { f.setVisible(true) } }
            .flatMap { IO { f.setFocusable(true) } }
            .flatMap { IO { f.setResizable(false) } }
            .flatMap { IO { f.validate() } }
    }

    fun updatePanelBackgroundColor(p: ColorChangerPanel): IO<Unit> {
        return randomColor().flatMap { c -> IO { p.setBackground(c) }.map { p.repaint() } }
    }

    fun initPanelAndWaitForKeyPress(width: Int, height: Int): IO<Unit> {
        return IO.just(JFrame("Press SPACE key to update background color")).flatMap { f ->
            IO.just(Dimension(width, height)).flatMap { d ->
                IO.just(ColorChangerPanel()).flatMap { p ->
                    IO { p.setPreferredSize(d) }.flatMap { initFrameProps(f, p) }
                }
            }
        }
    }

}